orchestrator/src/coordinator.rs

```diff
@@ -1,166 +1,295 @@
+        }
+
+        let execute_cargo = ExecuteCommandRequest {
+            cmd: "cargo".to_owned(),
+            args: args.into_iter().map(|s| s.to_owned()).collect(),
+            envs,
+            cwd: None,
+        };
+
+        let read_output = ReadFileRequest {
+            path: output_path.to_owned(),
+        };
+
+        let a = self.commander.one(write_main);
+        let b = self.commander.one(write_cargo_toml);
+
+        let (a, b) = join!(a, b);
+
+        // TODO: assert response success
+
+        let (stdout_tx, stdout_rx) = mpsc::channel(8);
+        let (stderr_tx, stderr_rx) = mpsc::channel(8);
+
+        let mut from_worker_rx = self.commander.many(execute_cargo).await;
+
+        let x = tokio::spawn({
+            let commander = self.commander.clone();
+            async move {
+                let mut success = false;
+
+                while let Some(container_msg) = from_worker_rx.recv().await {
+                    match container_msg {
+                        WorkerMessage::WriteFile(..) => todo!("nah"),
+                        WorkerMessage::ReadFile(..) => todo!("nah"),
+                        WorkerMessage::ExecuteCommand(..) => {
+                            success = true;
+                            break;
+                        }
+                        WorkerMessage::StdoutPacket(packet) => {
+                            stdout_tx.send(packet).await.ok(/* Receiver gone, that's OK */);
+                        }
+                        WorkerMessage::StderrPacket(packet) => {
+                            stderr_tx.send(packet).await.ok(/* Receiver gone, that's OK */);
+                        }
+                    }
+                }
+
+                let f: ReadFileResponse = commander.one(read_output).await;
+                let code = f.0;
+                let code = String::from_utf8(code).unwrap();
+
+                // TODO: Stop listening
+
+                Ok(CompileResponse2 { success, code })
+            }
+        });
+
+        Ok((x, stdout_rx, stderr_rx))
+    }
+}
+
+
+#[derive(Debug, Clone)]
+struct Commander {
+    to_worker_tx: mpsc::Sender<Multiplexed<CoordinatorMessage>>,
+    command_tx: mpsc::Sender<DemultiplexCommand>,
+    id: Arc<AtomicU64>,
+}
+
+impl Commander {
+    async fn demultiplex(
+        mut command_rx: mpsc::Receiver<DemultiplexCommand>,
+        mut rx: mpsc::Receiver<Multiplexed<WorkerMessage>>,
     ) {
         let mut waiting = HashMap::new();
+        let mut waiting_once = HashMap::new();
 
         loop {
             select! {
                 command = command_rx.recv() => {
                     let command = command.expect("Handle this");
                     match command {
-                        FanoutCommand::Listen(id, waiter) => {
+                        DemultiplexCommand::Listen(id, waiter) => {
                             waiting.insert(id, waiter);
                             // TODO: ensure not replacing
```

Comment:
Since each call(one/many) comes with a new id, replacing cannot happen.

Reply:
Exactly, so this todo will basically boil down to a panic that a replace did not happen mostly as a sanity check